contracts/0.8.25/utils/UpgradeTemplateV3.sol

```diff
@@ -0,0 +1,639 @@
+         || locator.accounting() != address(_accounting)
+         || locator.wstETH() != address(_wstETH)
+         || locator.vaultHub() != address(_vaultHub)
+         || locator.predepositGuarantee() != address(_predepositGuarantee)
+        ) {
+            revert IncorrectLocatorAddresses();
+        }
+    }
+
+    function _assertOldBurnerAllowances() internal view {
+        address oldBurner = address(_oldBurner);
+        uint256 infiniteAllowance = type(uint256).max;
+
+        if (_lido.allowance(_withdrawalQueue, oldBurner) != infiniteAllowance) {
+            revert IncorrectOldBurnerAllowance(_withdrawalQueue);
+        }
+
+        if (_lido.allowance(_simpleDvt, oldBurner) != infiniteAllowance) {
+            revert IncorrectOldBurnerAllowance(_simpleDvt);
+        }
+
+        if (_lido.allowance(_nodeOperatorsRegistry, oldBurner) != infiniteAllowance) {
+            revert IncorrectOldBurnerAllowance(_nodeOperatorsRegistry);
+        }
+
+        if (_lido.allowance(_csmAccounting, oldBurner) != infiniteAllowance) {
+            revert IncorrectOldBurnerAllowance(_csmAccounting);
+        }
+    }
+
+    function _assertProxyAdmin(IOssifiableProxy proxy, address admin) internal view {
+        if (proxy.proxy__getAdmin() != admin) revert IncorrectProxyAdmin(address(proxy));
+    }
+
+    function _assertProxyImplementation(IOssifiableProxy proxy, address implementation) internal view {
+        address actualImplementation = proxy.proxy__getImplementation();
+        if (actualImplementation != implementation) {
+            revert IncorrectProxyImplementation(address(proxy), actualImplementation);
+        }
+    }
+
+    function _assertOracleReportSanityCheckerRoles() internal view {
+        IOracleReportSanityChecker checker = _oracleReportSanityChecker;
+        _assertSingleOZRoleHolder(checker, DEFAULT_ADMIN_ROLE, _agent);
+        _assertZeroOZRoleHolders(checker, ALL_LIMITS_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, EXITED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, APPEARED_VALIDATORS_PER_DAY_LIMIT_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, ANNUAL_BALANCE_INCREASE_LIMIT_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, SHARE_RATE_DEVIATION_LIMIT_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, MAX_VALIDATOR_EXIT_REQUESTS_PER_REPORT_ROLE);
+        _assertZeroOZRoleHolders(checker, MAX_ITEMS_PER_EXTRA_DATA_TRANSACTION_ROLE);
+        _assertZeroOZRoleHolders(checker, MAX_NODE_OPERATORS_PER_EXTRA_DATA_ITEM_ROLE);
+        _assertZeroOZRoleHolders(checker, REQUEST_TIMESTAMP_MARGIN_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, MAX_POSITIVE_TOKEN_REBASE_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, SECOND_OPINION_MANAGER_ROLE);
+        _assertZeroOZRoleHolders(checker, INITIAL_SLASHING_AND_PENALTIES_MANAGER_ROLE);
+    }
+
+    function _assertInitialProxyImplementations() internal view {
```

Comment:
let's pin old impls for the following proxies:
- lido
- accounting oracle

we can cache them in the ctor reading from the `oldLocatorImpl` to public immutables